### PR TITLE
Personalized tagging

### PR DESCRIPTION
**Pull request by gyst**
This implements personalized tagging in a backward-compatible way. Documentation TODO.

The intervention consists of two steps:

1) Modify the subjects field in IDublinCore. Due to the way dexterity behaviors work that requires a wholesale fork of all of IDublinCore. This fork is then applied to all dexterity content types in the site. The modified behavior is backward compatible and gets/sets the context.subject attribute as usual (which is what is indexed into catalog/solr). Additionally however, the customized behavior _also_ shadows tagging activity into ploneintranet.network for use in step 2.

2) Activate a custom personalized vocabulary factory to power tag suggestions. This is done by overriding the @@getVocabulary view used by AJAX which resolves the vocabulary. We let our custom view resolve our custom vocabulary. The vocabulary is both (context, request) aware so aware of the current user. It then derives an optimal tag suggestion set based on the context and the user.

This stuff is set up in such a way that it's completely backward compatible and can be uninstalled without side effects.

### CODE REVIEW

**Comment by ghost** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ploneintranet.net/job/Plone.Intranet.Pull.Request.Builder/246/
Test FAILed.

**Comment by ghost** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ploneintranet.net/job/Plone.Intranet.Pull.Request.Builder/259/
Test PASSed.

File: src/ploneintranet/network/setuphandlers.bak [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+def replace_dublincore(fti, old, new, msg):
+    behaviors = []
+    for beh in fti.behaviors:
+        # plone.app.dexterity.behaviors.metadata.IDublinCore
+        if beh.startswith(old) and beh.endswith('IDublinCore'):
+            beh = beh.replace(old, new)
+            log.info('%s IDublinCore behavior on %s', msg, fti)
+        behaviors.append(beh)
+    fti.behaviors = behaviors
```
Comment (pilz) [inline src/ploneintranet/network/setuphandlers.bak]: .bak means we can remove it?

**Comment by pilz** [comment]
And that was also the only thing to say. The rest looks fine, nearly too harmless. Is there a good way to test this in my instance besides checking if it starts?

**Comment by ghost** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ploneintranet.net/job/Plone.Intranet.Pull.Request.Builder/270/
Test PASSed.

**Comment by gyst** [comment]
I just pushed a fix. Testing this is difficult. It only works outside of workspaces because of #317 - so it needs default Documents that require Barceloneta which means it needs #328. Best way to get that is on branch library. Additionally, checking the personalization cannot be done black-box, you really need to know how to work with multiple users on multiple docs to see the effects. See the tests.

To simply see it in action build branch library testcontent, goto http://cms.localhost:8080/Plone/library/human-resources/leave-policies/holidays/edit tab Categorization and try to set some tags. That exercises the personalized vocabulary.

**Comment by ghost** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ploneintranet.net/job/Plone.Intranet.Pull.Request.Builder/271/
Test PASSed.

**Comment by ghost** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ploneintranet.net/job/Plone.Intranet.Pull.Request.Builder/279/
Test PASSed.

**Comment by ghost** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ploneintranet.net/job/Plone.Intranet.Pull.Request.Builder/290/
Test FAILed.

**Comment by ghost** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ploneintranet.net/job/Plone.Intranet.Pull.Request.Builder/295/
Test FAILed.

**Comment by gyst** [comment]
Jenkins test this please.
That stupid docconv false positive again.

**Comment by mattss** [comment]
Retest this please jenkins

**Comment by mattss** [comment]
Retest this please jenkins

**Comment by gyst** [comment]
Jenkins please test this test please.

**Comment by ghost** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ploneintranet.net/job/Plone.Intranet.Pull.Request.Builder/299/
Test PASSed.

**Comment by gyst** [comment]
@pilz This is ready for merge and is a pre-requirement to get #343 working. If you can review and merge this, I can then review and merge #343 and #331

**Comment by gyst** [comment]
Actually, most of this work has already been merged as part of the library merge. What's left is a few polishing commits. Github doesn't show that but command line does.

[tag-auto-suggest] /app
app@thor(ploneintranet):app$ git log personalized-tagging ^master
commit fcafd29515644fac4c09dd94a51fcf6efdc4e581
Author: Guido A.J. Stevens guido.stevens@cosent.nl
Date:   Tue Jun 30 10:57:04 2015 +0000

```
tagging: add missing robot test wait
```

commit 1794e14d1b154a774d93e492ccd8ac76415b9b19
Author: Guido A.J. Stevens guido.stevens@cosent.nl
Date:   Tue Jun 30 08:24:45 2015 +0000

```
tagging: sort suggestions alphabetically
```

commit d1dee1ce7cdc8bea3c01a719218ff2aba0564d79
Author: Guido A.J. Stevens guido.stevens@cosent.nl
Date:   Tue Jun 30 08:13:14 2015 +0000

```
tagging: suggest existing tags you just removed, before saving
```

commit af88a1b37803be0df3c41da1b1a0caf9024dd8ec
Author: Guido A.J. Stevens guido.stevens@cosent.nl
Date:   Tue Jun 30 07:57:12 2015 +0000

```
tagging: handle coldstart case
```

commit eaf2be4e1bb5d79f4464daec3b263b2c2d15e0e5
Author: Guido A.J. Stevens guido.stevens@cosent.nl
Date:   Tue Jun 30 07:37:19 2015 +0000

```
make tagging test less fragile for changes in the tag space
```

[tag-auto-suggest] /app

File: src/ploneintranet/network/browser/configure.zcml [hunk 2]
```diff
@@ -30,4 +30,12 @@
       layer=".interfaces.IPloneIntranetNetworkLayer"
       />
 
+  <browser:page
+      name="getVocabulary"
```
Comment (mattss) [inline src/ploneintranet/network/browser/configure.zcml:34]: I'm slightly concerned about the generic name of this browser view, registered for *!

Comment (mattss) [inline src/ploneintranet/network/browser/configure.zcml:34]: Ah i see this is an override to an existing plone view, so necessary.

**Pull request merged**